docs: add starter for plugins with new plugin recipes (#22258)
gatsbyjs/gatsby: starters/gatsby-starter-plugin/README.md

<p align="center">
  <a href="https://www.gatsbyjs.org">
    <img alt="Gatsby" src="https://www.gatsbyjs.org/monogram.svg" width="60" />
  </a>
</p>
<h1 align="center">
  Starter for a Gatsby Plugin
</h1>

A minimal boilerplate for the essential files Gatsby looks for in a plugin.

## 🚀 Quick start

To get started creating a new plugin, you can follow these steps:

1. Initialize a new plugin from the starter with `gatsby new`

```
gatsby new my-plugin https://github.com/gatsbyjs/gatsby-starter-plugin
```

> If you already have a Gatsby site, you can use it. Otherwise, you can [create a new Gatsby site](https://www.gatsbyjs.org/tutorial/part-zero/#create-a-gatsby-site) to use your plugin in with the `gatsby new` command again.

Reviewer: ```suggestion
If you already have a Gatsby site, you can use it. Otherwise, you can [create a new Gatsby site](https://www.gatsbyjs.org/tutorial/part-zero/#create-a-gatsby-site) to test your plugin.
```

> Your directory structure might look like a little like this:

Reviewer: ```suggestion
Your directory structure will look similar to this:
```

```
/my-gatsby-site
├── gatsby-config.js
└── /src
    └── /pages
        └── /index.js
/my-plugin
├── gatsby-browser.js
├── gatsby-node.js
├── gatsby-ssr.js
├── index.js
├── package.json
└── README.md
```

With `my-gatsby-site` being your Gatsby site, and `my-plugin` being your plugin. You could also include the plugin in your [site's `plugins` folder](https://www.gatsbyjs.org/docs/loading-plugins-from-your-local-plugins-folder/).

2. Include the plugin in a Gatsby site

Inside of the `gatsby-config` file of your site (in this case, `my-gatsby-site`), include the plugin in the `plugins` array:

```
module.exports = {
  plugins: [
    // other gatsby plugins
    // ...
    require.resolve(`../my-plugin`)
  ],
}
```

The line `require.resolve('../my-plugin')` is what accesses the plugin based on its filepath on your computer, and adds it as a plugin when Gatsby runs.

> _You can use this method to test and develop your plugin before you publish it to a package registry like npm. Once it has been published you would instead install it and [add the plugin name to the array](https://www.gatsbyjs.org/docs/using-a-plugin-in-your-site/). You can read about other ways to connect your plugin to your site including using `npm link` or `yarn workspaces` in the [doc on creating local plugins](https://www.gatsbyjs.org/docs/creating-a-local-plugin/#developing-a-local-plugin-that-is-outside-your-project)._

Reviewer: ```suggestion
_You can use this method to test and develop your plugin before you publish it to a package registry like npm. Once published, you would instead install it and [add the plugin name to the array](https://www.gatsbyjs.org/docs/using-a-plugin-in-your-site/). You can read about other ways to connect your plugin to your site including using `npm link` or `yarn workspaces` in the [doc on creating local plugins](https://www.gatsbyjs.org/docs/creating-a-local-plugin/#developing-a-local-plugin-that-is-outside-your-project)._
```

3. Verify the plugin was added correctly

> The plugin added by the starter has one API implemented in the `gatsby-node` file that will log a message to the console when you run `gatsby develop` or `gatsby build` in the site that implements your plugin.

Reviewer: ```suggestion
The plugin added by the starter implements a single API in the `gatsby-node` that logs a message to the console. When you run `gatsby develop` or `gatsby build` in the site that implements your plugin, you should see this message.
```

> You can verify your plugin was added to your site correctly by running `gatsby develop` in the site.

Reviewer: ```suggestion
You can verify your plugin was added to your site correctly by running `gatsby develop` for the site.
```

You should now see a message logged to the console in the preinit phase of the Gatsby build process:

```
$ gatsby develop
success open and validate gatsby-configs - 0.033s
success load plugins - 0.074s
Loaded gatsby-plugin-minimal
success onPreInit - 0.016s
...
```

4. Rename the plugin in the `package.json`

When you clone the site, the information in the `package.json` will need to be updated. Name your plugin based off of [Gatsby's conventions for naming plugins](https://www.gatsbyjs.org/docs/naming-a-plugin/).

## 🧐 What's inside?

> These are the files and directories you'll see for a plugin using this starter. They are used because they are the same as the [files Gatsby looks for in plugins](https://www.gatsbyjs.org/docs/files-gatsby-looks-for-in-a-plugin/)]

Reviewer: ```suggestion
This starter generates the [files Gatsby looks for in plugins](https://www.gatsbyjs.org/docs/files-gatsby-looks-for-in-a-plugin/)].
```
